src/main/utility/sockaddr.rs

```diff
@@ -0,0 +1,688 @@
+// `libc::sockaddr_in6`
+const _: () = assert!(
+    std::mem::size_of::<libc::sockaddr_in6>()
+        == std::mem::size_of::<nix::sys::socket::SockaddrIn6>(),
+);
+
+impl SockaddrStorage {
+    /// SAFETY: The address must be fully initialized, including padding fields (for example
+    /// `sockaddr_in.sin_zero`), up until `len` bytes. Padding bytes do not need to be initialized.
+    /// The address does not need to be aligned.
+    pub unsafe fn from_ptr(
+        addr: *const MaybeUninit<u8>,
+        len: libc::socklen_t,
+    ) -> Option<SockaddrStorage> {
+        if addr.is_null() {
+            return None;
+        }
+
+        const STORAGE_LEN: usize = std::mem::size_of::<libc::sockaddr_storage>();
+
+        if (len as usize) > STORAGE_LEN {
+            return None;
+        }
+
+        // 'new_addr' starts will all bytes initialized
+        let mut new_addr = [MaybeUninit::new(0u8); STORAGE_LEN];
+
+        // after the copy, 'new_addr' may have uninitialized bytes if `addr` had padding bytes
+        unsafe { std::ptr::copy_nonoverlapping(addr, new_addr.as_mut_ptr(), len as usize) };
+
+        Some(SockaddrStorage {
+            addr: Addr { slice: new_addr },
+            len,
+        })
+    }
+
+    /// SAFETY: See [`Self::from_ptr`].
+    pub unsafe fn from_bytes(address: &[MaybeUninit<u8>]) -> Option<Self> {
+        unsafe { Self::from_ptr(address.as_ptr(), address.len().try_into().ok()?) }
+    }
+
+    /// Get the socket protocol family. Will return `None` if the socket address length is too
+    /// short, or if the family value does not correspond to a valid/known family.
+    pub fn family(&self) -> Option<AddressFamily> {
+        if (self.len as usize) < memoffset::span_of!(libc::sockaddr_storage, ss_family).end {
+            return None;
+        }
+
+        // SAFETY: we don't know what bytes of the address have initialized/uninitialized memory,
+        // but we should be guarenteed the the `ss_family` field is initialized for any socket type.
+        let family = unsafe { self.addr.storage }.ss_family;
+        AddressFamily::from_i32(family.into())
+    }
+
+    /// If the socket address represents a valid ipv4 socket address (correct family and length),
+    /// returns the ipv4 socket address.
+    pub fn as_inet(&self) -> Option<&nix::sys::socket::SockaddrIn> {
+        if !self.verify_fixed_size::<libc::sockaddr_in>(AddressFamily::Inet) {
+            return None;
+        }
+        // SAFETY: `nix::sys::socket::SockaddrIn` should be a transparent wrapper around a
+        // `libc::sockaddr_in`
+        Some(unsafe { &*(&self.addr.inet as *const _ as *const nix::sys::socket::SockaddrIn) })
+    }
+
+    /// If the socket address represents a valid ipv4 socket address (correct family and length),
+    /// returns the ipv4 socket address.
+    pub fn as_inet_mut(&mut self) -> Option<&mut nix::sys::socket::SockaddrIn> {
```

Comment:
I think uninitialized padding bytes are fine? The `SockaddrStorage::from_ptr` has the assumption that if the family field is a specific family (for example `AF_INET`), then we can safely cast to a `sockaddr_in`. It would be incorrect to create a `SockaddrStorage` from a `sockaddr_in` with `sin_family == AF_UNIX` for example.

Reply:
Yeah, we could probably do something like only return a mutable reference to other fields *not* including the family, or return some wrapper that prevents invalid mutations.... but agreed if we don't have a specific need then maybe better to just drop these for now or make them `unsafe`

Reply:
I removed the "mut" methods.